pkg/security/ptracer/proc.go

```diff
@@ -0,0 +1,286 @@
+// Copyright 2016-present Datadog, Inc.
+
+//go:build linux
+
+package ptracer
+
+import (
+	"bufio"
+	"context"
+	"fmt"
+	"io"
+	"os"
+	"path"
+	"slices"
+	"strings"
+	"time"
+
+	"github.com/DataDog/datadog-agent/pkg/security/proto/ebpfless"
+	"github.com/DataDog/gopsutil/process"
+)
+
+func isKThread(ppid, pid int32) bool {
+	return ppid == 2 || pid == 2
+}
+
+func listNotTracedProcess(tracerPID int32) (map[int32]*process.Process, error) {
+	pids, err := process.Pids()
+	if err != nil {
+		return nil, err
+	}
+
+	processes := make(map[int32]*process.Process)
+
+	for _, pid := range pids {
+		proc, err := process.NewProcess(pid)
+		if err != nil {
+			continue
+		}
+		processes[pid] = proc
+	}
+
+	var traced []int32
+
+	isTraced := func(proc *process.Process) bool {
+		for {
+			if proc.Pid == tracerPID || slices.Contains(traced, proc.Pid) {
+				return true
+			}
+
+			ppid, err := proc.Ppid()
+			if err != nil {
+				// if we can get the ppid ignore the process as it can be a race
+				return true
+			}
+
+			if ppid == 1 || ppid == 0 {
+				return false
+			}
+
+			// mark as traced kthreads so that we remove them
+			if isKThread(ppid, proc.Pid) {
+				return true
+			}
+
+			if proc = processes[ppid]; proc == nil {
+				return true
+			}
+		}
+	}
+
+	for pid, proc := range processes {
+		if isTraced(proc) {
+			traced = append(traced, pid)
+		}
+	}
+
+	// remove traced from the process list
+	for _, pid := range traced {
+		delete(processes, pid)
+	}
+
+	return processes, nil
+}
+
+func pidTTY(pid int32) string {
+	fdPath := fmt.Sprintf("/proc/%d/fd/0", pid)
+
+	ttyPath, err := os.Readlink(fdPath)
+	if err != nil {
+		return ""
+	}
+
+	if ttyPath == "/dev/null" {
+		return ""
+	}
+
+	if strings.HasPrefix(ttyPath, "/dev/pts") {
+		return "pts" + path.Base(ttyPath)
+	}
+
+	if strings.HasPrefix(ttyPath, "/dev") {
+		return path.Base(ttyPath)
+	}
+
+	return ""
+}
+
+// MaxEnvVarsCollected max env to be collected
+const MaxEnvVarsCollected = 256
+
+func zeroSplitter(data []byte, atEOF bool) (advance int, token []byte, err error) {
+	for i := 0; i < len(data); i++ {
+		if data[i] == '\x00' {
+			return i + 1, data[:i], nil
+		}
+	}
+	if !atEOF {
+		return 0, nil, nil
+	}
+	return 0, data, bufio.ErrFinalToken
+}
+
+func newEnvScanner(f *os.File) (*bufio.Scanner, error) {
+	_, err := f.Seek(0, io.SeekStart)
+	if err != nil {
+		return nil, err
+	}
+
+	scanner := bufio.NewScanner(f)
+	scanner.Split(zeroSplitter)
+
+	return scanner, nil
+}
+
+func envVars(pid int32) []string {
+	filename := fmt.Sprintf("/proc/%d/environ", pid)
+
+	f, err := os.Open(filename)
+	if err != nil {
+		return nil
+	}
+	defer f.Close()
+
+	// first pass collecting only priority variables
+	scanner, err := newEnvScanner(f)
+	if err != nil {
+		return nil
+	}
+
+	var envs []string
+	envCounter := 0
+
+	for scanner.Scan() {
+		text := scanner.Text()
+		if len(text) > 0 {
+			envs = append(envs, text)
+
+			envCounter++
+			if envCounter > MaxEnvVarsCollected {
+				break
+			}
```

Comment:
We could use the `truncateEnvs()` to prioritize/truncate envs after scanning them?